In [1]:
import pandas as pd


In [2]:
df = pd.read_csv('/content/RECOMMENDATION SYSTEM USING HUGGING FACES TRANSFORMERS/anime.csv')
df

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members  
0        9.37   200630  
1        9.26   793665  
2        9.25   114262  
3        9.17   673572  
4        9.16   151266  
...       ...      ...  
12289    4.15      211  
12290    4.28      183  
12291    4.88      219  
12292    4.98      175  
12293    5.46      142  

[12294 rows x 7 columns]

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12294 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12294 non-null  int64  
 1   name      12294 non-null  object 
 2   genre     12232 non-null  object 
 3   type      12269 non-null  object 
 4   episodes  12294 non-null  object 
 5   rating    12064 non-null  float64
 6   members   12294 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 672.5+ KB


In [ ]:
# here we have null values in genre  so for that we dropping the null values

In [4]:
df = df.dropna()

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 12017 entries, 0 to 12293
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   anime_id  12017 non-null  int64  
 1   name      12017 non-null  object 
 2   genre     12017 non-null  object 
 3   type      12017 non-null  object 
 4   episodes  12017 non-null  object 
 5   rating    12017 non-null  float64
 6   members   12017 non-null  int64  
dtypes: float64(1), int64(2), object(4)
memory usage: 751.1+ KB


In [8]:
df["description"] = df["name"] + " " + df["genre"] + " " + df["type"] + " " + "rating : " + df["rating"].astype(str) + " "+ "episodes : " + df["episodes"].astype(str)


<ipython-input-8-22faa2d4f414>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["description"] = df["name"] + " " + df["genre"] + " " + df["type"] + " " + "rating : " + df["rating"].astype(str) + " "+ "episodes : " + df["episodes"].astype(str)


here we try to enter description for the the movies by combining all the columns like name, genre, type, rating, episodes

In [9]:
df

anime_id                                               name  \
0         32281                                     Kimi no Na wa.   
1          5114                   Fullmetal Alchemist: Brotherhood   
2         28977                                           Gintama°   
3          9253                                        Steins;Gate   
4          9969                                      Gintama&#039;   
...         ...                                                ...   
12289      9316       Toushindai My Lover: Minami tai Mecha-Minami   
12290      5543                                        Under World   
12291      5621                     Violence Gekiga David no Hoshi   
12292      6133  Violence Gekiga Shin David no Hoshi: Inma Dens...   
12293     26081                   Yasuji no Pornorama: Yacchimae!!   

                                                   genre   type episodes  \
0                   Drama, Romance, School, Supernatural  Movie        1   
1      Action, Adventure, Drama, Fantasy, Magic, Mili...     TV       64   
2      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
3                                       Sci-Fi, Thriller     TV       24   
4      Action, Comedy, Historical, Parody, Samurai, S...     TV       51   
...                                                  ...    ...      ...   
12289                                             Hentai    OVA        1   
12290                                             Hentai    OVA        1   
12291                                             Hentai    OVA        4   
12292                                             Hentai    OVA        1   
12293                                             Hentai  Movie        1   

       rating  members                                        description  
0        9.37   200630  Kimi no Na wa. Drama, Romance, School, Superna...  
1        9.26   793665  Fullmetal Alchemist: Brotherhood Action, Adven...  
2        9.25   114262  Gintama° Action, Comedy, Historical, Parody, S...  
3        9.17   673572  Steins;Gate Sci-Fi, Thriller TV rating : 9.17 ...  
4        9.16   151266  Gintama&#039; Action, Comedy, Historical, Paro...  
...       ...      ...                                                ...  
12289    4.15      211  Toushindai My Lover: Minami tai Mecha-Minami H...  
12290    4.28      183  Under World Hentai OVA rating : 4.28 episodes : 1  
12291    4.88      219  Violence Gekiga David no Hoshi Hentai OVA rati...  
12292    4.98      175  Violence Gekiga Shin David no Hoshi: Inma Dens...  
12293    5.46      142  Yasuji no Pornorama: Yacchimae!! Hentai Movie ...  

[12017 rows x 8 columns]

In [10]:
from transformers import AutoTokenizer, AutoModel
import torch
import torch.nn.functional as F
tokenizer = AutoTokenizer.from_pretrained('sentence-transformers/all-MiniLM-L6-v2')
model = AutoModel.from_pretrained('sentence-transformers/all-MiniLM-L6-v2')

/usr/local/lib/python3.11/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

In [11]:
sentences = [" some great movie","another funny movie"]
encoded_input = tokenizer(sentences, padding=True, truncation=True, return_tensors="pt")
encoded_input

{'input_ids': tensor([[ 101, 2070, 2307, 3185,  102],
        [ 101, 2178, 6057, 3185,  102]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]])}

we use **padding=True **

cause In machine learning, particularly when working with text data and neural networks, "padding = True" is used to ensure that all input sequences in a batch have the same length by adding extra "padding tokens" to shorter sequences, allowing the model to process data consistently without issues arising from varying input lengths; essentially, it prevents data loss at the edges by adding extra space around the actual data

we use **truncation = true cause**
True or 'longest_first' : truncate to a maximum length specified by the max_length argument or the maximum length accepted by the model if no max_length is provided ( max_length=None ). This will truncate token by token, removing a token from the longest sequence in the pair until the proper length is reached
make sure it token length  remains in the specified no of tokens it accepts
 ex:- bert models has max length of tokens as 520
   

In [14]:
with torch.no_grad():
    model_output = model(**encoded_input)
model_output
# here we use pre trained models so no need of any gradient so we sue no grad here
# it has coverted our example sentences into tokens and that tokens  into embeddings


BaseModelOutputWithPoolingAndCrossAttentions(last_hidden_state=tensor([[[-0.1721, -0.2529, -0.0292,  ..., -0.1430, -0.2251, -0.1713],
         [-1.5610,  0.2194,  0.3281,  ...,  0.2852,  0.4855,  1.7615],
         [-0.5192,  0.6054, -0.0881,  ...,  0.1461, -0.4203,  0.1676],
         [-0.5308,  0.1670, -0.5925,  ..., -0.6870,  0.4860, -0.2893],
         [-0.7257, -0.2807, -0.8826,  ..., -0.3162, -0.0586, -0.5847]],

        [[-0.3227, -0.3661, -0.1568,  ..., -0.1735, -0.0871, -0.0466],
         [-1.1943, -0.0665, -0.5749,  ..., -0.2577,  1.2235,  1.8534],
         [-0.4567, -0.3737, -0.4876,  ...,  0.2858,  0.5438, -0.2028],
         [-0.7948,  0.0621, -0.8832,  ..., -0.6371,  0.9120, -0.0876],
         [-0.7753,  0.1839, -0.8327,  ..., -0.4386, -0.0675, -0.1969]]]), pooler_output=tensor([[-5.6678e-02,  4.2709e-02,  2.0903e-02,  5.9918e-02, -5.9641e-02,
          4.6265e-03,  2.4511e-02,  4.7126e-02, -6.1902e-02, -7.9068e-03,
         -3.2739e-04,  6.6537e-02,  3.7929e-02, -1.6592e-02,

[ 101, 2070, 2307, 3185,  102] =  some great movie (tokens for this sentence)

  [ 101, 2178, 6057, 3185,  102]] = another funny movie"(tokens for this sentences)

In [ ]:
# # [-0.1721, -0.2529, -0.0292,  ..., -0.1430, -0.2251, -0.1713],
#          [-1.5610,  0.2194,  0.3281,  ...,  0.2852,  0.4855,  1.7615],
#          [-0.5192,  0.6054, -0.0881,  ...,  0.1461, -0.4203,  0.1676],
#          [-0.5308,  0.1670, -0.5925,  ..., -0.6870,  0.4860, -0.2893],
#         [-0.7257, -0.2807, -0.8826,  ..., -0.3162, -0.0586, -0.5847]
# here we are summing it up and dividing it with valid tokens with mean pooling

In [15]:
def mean_pooling(model_output, attention_mask):
    token_embeddings = model_output[0] #First element of model_output contains all token
    input_mask_expanded = attention_mask.unsqueeze(-1).expand(token_embeddings.size()).float()
    return torch.sum(token_embeddings * input_mask_expanded, 1) / torch.clamp(input_mask_expanded.sum(1), min=1e-9)

In [16]:
sentence_embeddings = mean_pooling(model_output, encoded_input["attention_mask"])
sentence_embeddings

tensor([[-7.0177e-01,  9.1624e-02, -2.5286e-01,  2.4189e-01, -2.0731e-01,
          2.2241e-01,  4.5603e-01, -2.7595e-01,  7.9938e-01, -2.8093e-01,
          2.6236e-01,  1.0508e-01, -2.4485e-01,  1.5632e-01, -2.1632e-01,
         -3.5087e-02,  2.5945e-01,  4.6815e-02, -1.8705e-01, -5.6078e-01,
         -3.4497e-01,  2.5035e-01,  2.6238e-01,  3.1524e-01, -1.0997e+00,
          1.4417e-01,  1.7577e-01,  9.6571e-03, -6.1394e-01, -2.2132e-01,
         -2.1643e-01,  4.3529e-01, -6.2619e-01,  2.0323e-01, -2.4971e-01,
          7.4963e-02,  2.2987e-01,  1.6839e-02, -1.8499e-01, -1.3225e-01,
         -2.2322e-01, -5.9756e-01,  6.4447e-01, -8.7913e-02,  2.3552e-01,
         -2.0751e-01, -1.5824e-01, -9.5676e-02,  3.2696e-01,  2.5714e-01,
         -1.9483e-01,  5.5819e-01, -3.5864e-01, -5.2131e-02,  3.3138e-01,
         -2.8594e-01, -3.9661e-01, -6.5131e-01,  2.3990e-01, -9.6422e-01,
          4.3170e-01, -2.5392e-01,  1.9183e-01, -3.7302e-01,  5.0330e-01,
          1.4846e-01,  8.3340e-02,  1.

# An attention mask is used in transformer models to control which parts of an input sequence the model should focus on during calculations, essentially filtering out irrelevant information like padding tokens, and ensuring the model only attends to the meaningful parts of the sequence, particularly when dealing with variable-length inputs or maintaining causal relationships in autoregressive models.

In [22]:
pip install -U sentence-transformers


In [25]:
from sentence_transformers import SentenceTransformer
model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
anime_embeddings = model.encode(df['description'].tolist())

In [26]:
anime_embeddings

array([[-0.03336235, -0.02177108,  0.04331539, ..., -0.04460806,
        -0.11355214, -0.02200367],
       [-0.03578721, -0.05518353,  0.00180863, ...,  0.00709238,
        -0.05171949, -0.037193  ],
       [-0.10148681, -0.04010534, -0.01179383, ..., -0.02266363,
        -0.08907078,  0.02672056],
       ...,
       [ 0.04307306,  0.06884463, -0.02271423, ..., -0.05767418,
        -0.01226672,  0.00805722],
       [ 0.00284595,  0.06014551, -0.01945104, ..., -0.06226813,
        -0.01299132,  0.0355329 ],
       [ 0.01507893, -0.04404013,  0.03074887, ..., -0.01561166,
         0.00055608, -0.01484828]], dtype=float32)

In [39]:
query = "funny movies to watch"


In [40]:
from sklearn.metrics.pairwise import cosine_similarity
def get_recommendations(query, top_k=5):
  query_embedding = model.encode([query])
  similarities = cosine_similarity(query_embedding, anime_embeddings)
  top_indices = similarities[0].argsort()[-top_k:][::-1]
  return df.iloc[top_indices]

In [41]:
recommendations = get_recommendations(query, 5)
recommendations

anime_id                   name  \
7191     29515         Happy ComeCome   
2672      1796  Dirty Pair: The Movie   
225       5681            Summer Wars   
1686      1764      Slam Dunk (Movie)   
5723     16149             Kick-Heart   

                                                  genre   type episodes  \
7191                      Comedy, Sci-Fi, Slice of Life  Movie        1   
2672                  Adventure, Comedy, Police, Sci-Fi  Movie        1   
225                                      Comedy, Sci-Fi  Movie        1   
1686  Comedy, Drama, School, Shounen, Slice of Life,...  Movie        1   
5723                            Comedy, Romance, Sports  Movie        1   

      rating  members                                        description  
7191    5.65      758  Happy ComeCome Comedy, Sci-Fi, Slice of Life M...  
2672    7.19     3227  Dirty Pair: The Movie Adventure, Comedy, Polic...  
225     8.30   223359  Summer Wars Comedy, Sci-Fi Movie rating : 8.3 ...  
1686    7.47     8750  Slam Dunk (Movie) Comedy, Drama, School, Shoun...  
5723    6.37     8007  Kick-Heart Comedy, Romance, Sports Movie ratin...